cranelift-faerie/src/backend.rs

```diff
@@ -74,9 +79,164 @@ pub struct FaerieBackend {
+
+        let mut encoder = CFIEncoder::new(&reg_mapper);
+
+        for inst in initial_state
+            .iter()
+            .flat_map(|change| encoder.translate(change))
+        {
+            cie.add_instruction(inst);
+        }
+
+        let cie_id = self.table.add_cie(cie);
+
+        (cie_id, encoder)
+    }
+
+    pub fn new() -> FrameSink {
+        FrameSink {
+            fn_names: vec![],
+            table: gimli::write::FrameTable::default(),
+        }
+    }
+
+    pub fn address_for(&mut self, name: &str) -> Address {
+        // adding a FrameDescriptionEntry for a function twice would be a bug,
+        // so we can confidently expect that `name` will not be provided more than once.
+        // So `name` is always new, meaning we can just add it and return its index
+        self.fn_names.push(name.to_string());
+        Address::Symbol {
+            symbol: self.fn_names.len() - 1,
+            addend: 0,
+        }
+    }
+
+    /// Add a FrameDescriptionEntry to the FrameTable we're constructing
+    ///
+    /// This will always use the default CIE (which was build with this `FrameSink`).
+    pub fn add_fde(&mut self, cie_id: gimli::write::CieId, fd_entry: FrameDescriptionEntry) {
+        self.table.add_fde(cie_id, fd_entry);
+    }
+}
+
+struct FaerieDebugSink<'a> {
+    pub data: &'a mut Vec<u8>,
+    pub functions: &'a [String],
+    pub artifact: &'a mut faerie::Artifact,
+}
+
+impl<'a> gimli::write::Writer for FaerieDebugSink<'a> {
+    type Endian = gimli::LittleEndian;
+
+    fn endian(&self) -> Self::Endian {
+        gimli::LittleEndian
+    }
+    fn len(&self) -> usize {
+        self.data.len()
+    }
+    fn write(&mut self, bytes: &[u8]) -> gimli::write::Result<()> {
+        self.data.extend_from_slice(bytes);
+        Ok(())
+    }
+
+    fn write_at(&mut self, offset: usize, bytes: &[u8]) -> gimli::write::Result<()> {
+        if offset + bytes.len() > self.data.len() {
+            return Err(gimli::write::Error::LengthOutOfBounds);
+        }
+        self.data[offset..][..bytes.len()].copy_from_slice(bytes);
+        Ok(())
+    }
+
+    fn write_eh_pointer(
+        &mut self,
+        address: Address,
+        eh_pe: gimli::DwEhPe,
+        size: u8,
+    ) -> gimli::write::Result<()> {
+        // we only support PC-relative 4byte signed offsets for eh_frame pointers currently. Other
+        // encodings may be permissible, but aren't seen even by gcc/clang/etc, and have not been
+        // tested. Currently, relocations used for addresses expect to be relocating four bytes,
+        // PC-relative, and larger pointer sizes would require selection of other relocation types.
+        assert!(eh_pe.0 == 0x1b);
+
+        // if size is not 4, then the size indicated by `eh_pe` doesn't match with the pointer
+        // we're trying to encode. That's a logical bug, possibly in gimli?
+        assert!(size == 4);
+
+        self.write_address(address, size)
```

Comment:
And so this should be passing a fixed 4, since we've already checked that this is using `DW_EH_PE_sdata4`.

Reply:
Also, this shouldn't call `write_address`, since `write_address` must only be used for absolute pointers. If you want to share code between the two then put it in another function.

Reply:
I didn't particularly care about sharing code, I just misunderstood which of `write_eh_pointer` and `write_address` were supposed to do what.